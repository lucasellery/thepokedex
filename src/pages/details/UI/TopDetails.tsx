import React from 'react';
import { View, Text, Image, StyleSheet } from 'react-native';
import addZeros from '../../../utils/addZeros';

type TopDetailsProps = {
  pokedexEntry: number;
  pokemonName: string;
  pokemonTypes:string[];
  pokemonImage: string;
  pokemonTypeColor: {
    main: string;
  }
}

const TopDetails: React.FC<TopDetailsProps> = ({ pokedexEntry, pokemonName, pokemonTypes, pokemonImage, pokemonTypeColor }) => {
  return (
    <View testID="top-details" style={[styles.container, { backgroundColor: pokemonTypeColor?.main }]}>
      <View style={styles.wrapper}>
        <View>
          <Text style={styles.entry} testID="pokedex-entry-number">
            {`#${addZeros(pokedexEntry)}`}
          </Text>
          <Text style={styles.pokemonName} testID="pokemon-name">
            {pokemonName}
          </Text>
        </View>

        <View style={styles.badgesWrapper}>
          {pokemonTypes?.map(type => <Text style={styles.badge} key={type} testID="pokemon-type">{type}</Text>)}
        </View>
      </View>

        <Image style={styles.pokemonImage} testID="pokemon-image" source={{ uri: pokemonImage }}/>
    </View>
  );
}

export default TopDetails;

const styles = StyleSheet.create({
  container: {
    padding: 20,
    paddingBottom: 100
  },
  wrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  entry: {
    color: 'white',
    fontSize: 16,
    fontWeight: 'bold',
    marginBottom: 10
  },
  pokemonName: {
    fontSize: 32,
    fontWeight: 'bold',
  },
  pokemonImage: {
    width: 120,
    height: 120,
    position: 'absolute',
    right: 0,
    top: 35
  },
  badge: {
    paddingHorizontal: 10,
    paddingVertical: 5,
    backgroundColor: 'rgba(0, 0, 0, 0.1)',
    borderRadius: 20,
    color: '#FFFFFF',
    fontSize: 16,
    fontWeight: 'bold',
    marginHorizontal: 2.5
  },
  badgesWrapper: {
    flexDirection: 'row',
    alignSelf: 'flex-start'
  }
});