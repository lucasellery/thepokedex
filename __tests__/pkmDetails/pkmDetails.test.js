import React from 'react'
import { render } from '@testing-library/react-native' 
import TopDetails from '../../src/pages/details/UI/TopDetails';

describe('Pokemon details test', () => {
  
  it('Should render TopDetails correctly', () => {
    const { queryByTestId } = render(<TopDetails />)
    const component = queryByTestId('top-details')
    expect(component).toBeTruthy()
  })

  it('should has a pkm pokedex position', () => {
    const { queryByTestId } = render(<TopDetails pokedexEntry={25} />)
    const component = queryByTestId('pokedex-entry-number')
    expect(component.props.children).toBe('#025')
  })

  it('should render pkm name', () => {
    const pokemonName = 'pikachu'

    const { queryByTestId } = render(<TopDetails pokemonName={pokemonName} />)
    const component = queryByTestId('pokemon-name')
    expect(component.props.children).toBe(pokemonName)
  })

  it('should render pkm type badge', () => {
    const pokemonTypes = ['electric', 'normal']

    const { queryAllByTestId } = render(<TopDetails pokemonTypes={pokemonTypes} />)
    const components = queryAllByTestId('pokemon-type')
    
    expect(components.length).toEqual(pokemonTypes.length);

    components.forEach((component, index) => {
      expect(component.props.children).toBe(pokemonTypes[index])
    })
  })

  it('should render pkm image', () => {
    const url = 'url'

    const { queryByTestId } = render(<TopDetails pokemonImage={url} />)
    const component = queryByTestId('pokemon-image')
    expect(component.props.source.uri).toBe(url);
  })

  it('should has a background color prop', () => {
    const pokemonTypeColor = {main: '#1234'};

    const { queryByTestId } = render(<TopDetails pokemonTypeColor={pokemonTypeColor} />)
    const component = queryByTestId('top-details')
    expect(component.props.style[1].backgroundColor).toBe(pokemonTypeColor?.main);
  })

  it('should match snapshot <TopDetails />', () => {
    const pokemonName = 'pikachu'
    const pokedexEntry = 32
    const pokemonTypes = ['electric', 'normal']
    const url = 'url'

    const component = render(<TopDetails 
      pokemonImage={url}
      pokemonName={pokemonName} 
      pokemonTypes={pokemonTypes} 
      pokedexEntry={pokedexEntry} />).toJSON()
    expect(component).toMatchSnapshot()
  })
});