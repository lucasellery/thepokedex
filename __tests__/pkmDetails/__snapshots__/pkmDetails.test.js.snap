// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon details test should match snapshot <TopDetails /> 1`] = `
<View
  style={
    Array [
      Object {
        "padding": 20,
        "paddingBottom": 100,
      },
      Object {
        "backgroundColor": undefined,
      },
    ]
  }
  testID="top-details"
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View>
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
            "marginBottom": 10,
          }
        }
        testID="pokedex-entry-number"
      >
        #032
      </Text>
      <Text
        style={
          Object {
            "fontSize": 32,
            "fontWeight": "bold",
          }
        }
        testID="pokemon-name"
      >
        pikachu
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "flex-start",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.1)",
            "borderRadius": 20,
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 2.5,
            "paddingHorizontal": 10,
            "paddingVertical": 5,
          }
        }
        testID="pokemon-type"
      >
        electric
      </Text>
      <Text
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.1)",
            "borderRadius": 20,
            "color": "#FFFFFF",
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 2.5,
            "paddingHorizontal": 10,
            "paddingVertical": 5,
          }
        }
        testID="pokemon-type"
      >
        normal
      </Text>
    </View>
  </View>
  <Image
    source={
      Object {
        "uri": "url",
      }
    }
    style={
      Object {
        "height": 120,
        "position": "absolute",
        "right": 0,
        "top": 35,
        "width": 120,
      }
    }
    testID="pokemon-image"
  />
</View>
`;
